### PR TITLE
[mypy] Annotate other/lru_cache and other/lfu_cache

### PR DESCRIPTION
**Pull request by spazm**
### Describe your change:

Annotation and improvements for lru_cache and lfu_cache.  
These contain a lot of duplicated code so I have put them into a single PR.

+ Annotates existing lru_cache algorithm
+ Adds tests for the internal double linked list
+ Expands algorithm annotations to be generic of key and value types
+ Added __repr__ to enable clean doctests and clarity in the repl
+ mypy required using `setattr` rather than creating an attr through assignment.

+ Makes analogous changes to lfu_cache.

PR is broken into multiple commits each handling a logical chunk and is meant to be clear when viewing each commit individually.

#### Before - lru_cache
```
% git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

 % mypy --ignore-missing-imports --install-types lru_cache.py      :(-(master)-~/src/github/TheAlogorithms/Python/other
lru_cache.py:44: error: "None" has no attribute "next"
lru_cache.py:44: error: "None" has no attribute "prev"
lru_cache.py:94: error: Need type annotation for "decorator_function_to_instance_map" (hint: "decorator_function_to_instance_map: Dict[<type>, <type>] = ...")
lru_cache.py:102: error: Need type annotation for "cache" (hint: "cache: Dict[<type>, <type>] = ...")
lru_cache.py:185: error: "Callable[[VarArg(Any), KwArg(Any)], Any]" has no attribute "cache_info"
Found 5 errors in 1 file (checked 1 source file)
```
#### After -- lru_cache
```
% git checkout master 
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

% mypy --ignore-missing-imports --install-types lru_cache.py
Success: no issues found in 1 source file

% mypy --strict lru_cache.py
Success: no issues found in 1 source file
```

#### Before -- lfu_cache
```
 % git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

% mypy --ignore-missing-imports --install-types lfu_cache.py
lfu_cache.py:42: error: "None" has no attribute "key"
lfu_cache.py:42: error: "None" has no attribute "freq"
lfu_cache.py:44: error: "None" has no attribute "next"
lfu_cache.py:44: error: "None" has no attribute "prev"
lfu_cache.py:45: error: "None" has no attribute "prev"
lfu_cache.py:54: error: "None" has no attribute "next"
lfu_cache.py:54: error: "None" has no attribute "prev"
lfu_cache.py:92: error: Need type annotation for "decorator_function_to_instance_map" (hint: "decorator_function_to_instance_map: Dict[<type>, <type>] = ...")
lfu_cache.py:100: error: Need type annotation for "cache" (hint: "cache: Dict[<type>, <type>] = ...")
lfu_cache.py:179: error: "Callable[[VarArg(Any), KwArg(Any)], Any]" has no attribute "cache_info"
Found 10 errors in 1 file (checked 1 source file)
```

#### After -- lfu_cache
```
% git checkout -
Switched to branch 'mypy-fix-other-lru_cache'
Your branch is up to date with 'origin/mypy-fix-other-lru_cache'.

% mypy --ignore-missing-imports --install-types lfu_cache.py
Success: no issues found in 1 source file

% mypy --strict lfu_cache.py
Success: no issues found in 1 source file
```

Related: #4058

* [ ] Add an algorithm?
* [X] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes ~~one~~ two algorithm files.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: other/lfu_cache.py [hunk 1]
```diff
@@ -110,73 +216,94 @@ def __repr__(self) -> str:
...
             node = self.list.remove(self.cache[key])
+            assert node is not None  # node guaranteed to be in list
             node.val = value
             self.list.add(node)
 
-    @staticmethod
-    def decorator(size: int = 128):
+    @classmethod
+    def decorator(
+        cls, size: int = 128
```
Comment (poyea) [inline other/lfu_cache.py:284]: What's `cls`?

Comment (poyea) [reply other/lfu_cache.py:284]: Is it the `class` or `class_object`?

File: other/lfu_cache.py [hunk 2]
```diff
@@ -110,73 +216,94 @@ def __repr__(self) -> str:
...
+                # delete first node when over capacity
+                first_node = self.list.head.next
+
+                # guaranteed to have a non-None first node when num_keys > 0
+                # explain to type checker via assertions
+                assert first_node is not None
+                assert first_node.key is not None
+                assert (
+                    self.list.remove(first_node) is not None
+                )  # node guaranteed to be in list assert node.key is not None
```
Comment (cclauss) [inline other/lfu_cache.py:268]: Please do not wrap lines just to add comment...
```suggestion
                # node guaranteed to be in list assert node.key is not None
                assert self.list.remove(first_node) is not None
```

Also, please remove these files from `mypy.ini` so the tests run.

File: other/lru_cache.py [hunk 3]
```diff
@@ -1,72 +1,190 @@
...
+U = TypeVar("U")
 
-class DoubleLinkedListNode:
+
+class DoubleLinkedListNode(Generic[T, U]):
     """
     Double Linked List Node built specifically for LRU Cache
+
+    >>> node = DoubleLinkedListNode(1,1)
+    >>> node
```
Comment (cclauss) [inline other/lru_cache.py:14]:
```suggestion
    >>> DoubleLinkedListNode(1,1)
```

File: other/lru_cache.py [hunk 4]
```diff
@@ -1,72 +1,190 @@
...
+    >>> cache.list
+    DoubleLinkedList,
+        Node: key: None, val: None, has next: True, has prev: False,
+        Node: key: 2, val: 2, has next: True, has prev: True,
+        Node: key: 1, val: 1, has next: True, has prev: True,
+        Node: key: None, val: None, has next: False, has prev: True
+
+    >>> cache.cache
+    {1: Node: key: 1, val: 1, has next: True, has prev: True, \
+2: Node: key: 2, val: 2, has next: True, has prev: True}
```
Comment (cclauss) [inline other/lru_cache.py:166]:
```suggestion
    >>> cache.cache  # doctest: +NORMALIZE_WHITESPACE
    {1: Node: key: 1, val: 1, has next: True, has prev: True,
     2: Node: key: 2, val: 2, has next: True, has prev: True}
```

File: other/lru_cache.py [hunk 5]
```diff
@@ -1,72 +1,190 @@
...
+    >>> cache.list
+    DoubleLinkedList,
+        Node: key: None, val: None, has next: True, has prev: False,
+        Node: key: 1, val: 1, has next: True, has prev: True,
+        Node: key: 3, val: 3, has next: True, has prev: True,
+        Node: key: None, val: None, has next: False, has prev: True
+
+    >>> cache.cache
+    {1: Node: key: 1, val: 1, has next: True, has prev: True, \
+3: Node: key: 3, val: 3, has next: True, has prev: True}
```
Comment (cclauss) [inline other/lru_cache.py:179]:
```suggestion
    >>> cache.cache  # doctest: +NORMALIZE_WHITESPACE
    {1: Node: key: 1, val: 1, has next: True, has prev: True,
     3: Node: key: 3, val: 3, has next: True, has prev: True}
```

File: other/lru_cache.py [hunk 6]
```diff
@@ -127,62 +245,81 @@ def __contains__(self, key: int) -> bool:
 
         return key in self.cache
 
-    def get(self, key: int) -> int | None:
+    def get(self, key: T) -> U | None:
         """
-        Returns the value for the input key and updates the Double Linked List. Returns
-        None if key is not present in cache
+        Returns the value for the input key and updates the Double Linked List.
+        Returns None if key is not present in cache
```
Comment (cclauss) [inline other/lru_cache.py:251]: I would thrown a `KeyError` but we will go with it as is.
```suggestion
        Returns the value for the input key and updates the Double Linked List
        or returns None if key is not present in cache.
```

File: other/lru_cache.py (hunk 4 above)
Comment (spazm) [reply other/lru_cache.py:166]: Cool.  I did not know about `+NORMALIZE_WHITESPACE` and am excited to have learned it.

File: other/lru_cache.py (hunk 6 above)
Comment (spazm) [reply other/lru_cache.py:251]: I agree that is a better interface.  I just documented it while figuring out which types things /actually/ use.  :)

File: other/lfu_cache.py (hunk 1 above)
Comment (spazm) [reply other/lfu_cache.py:284]: It is the class,  a `Type[LFUCache[T, U]`.

File: other/lru_cache.py (hunk 3 above)
Comment (spazm) [reply other/lru_cache.py:14]: done

File: other/lru_cache.py (hunk 4 above)
Comment (spazm) [reply other/lru_cache.py:166]: done

File: other/lru_cache.py (hunk 5 above)
Comment (spazm) [reply other/lru_cache.py:179]: done

File: other/lfu_cache.py (hunk 2 above)
Comment (spazm) [reply other/lfu_cache.py]: merge upstream/master in and updated mypy.ini to remove lfu_cache and lru_cache.
Fixed comment and removed broken lines.

File: other/lru_cache.py (hunk 6 above)
Comment (spazm) [reply other/lru_cache.py:250]: Added a note that throwing KeyError would be a better and more python interface.

**Review by cclauss — APPROVED** [review]
Nice!!

**Pull request merged**